enhancements/machine-api/control-plane-machineset-vertical-scaling.md

```diff
@@ -0,0 +1,348 @@
+---
+title: control-plane-machineset-vertical-scaling
+authors:
+  - "@bergmannf"
+reviewers:
+  - @"JoelSpeed"
+approvers:
+  - @"JoelSpeed"
+api-approvers: 
+  - None
+creation-date: 2023-10-30
+last-updated: 2023-10-30
+tracking-link:
+  - https://issues.redhat.com/browse/OSD-15261
+see-also: []
+replaces: []
+superseded-by: []
+---
+
+# Control plane machineset automatic vertical scaling
+
+## Summary
+
+To allow control plane nodes to remain at an adequate size for its cluster, this
+proposal introduces new configuration for the control plane machineset operator
+to allow it to make automated scaling decisions for control plane node sizing.
+
+## Motivation
+
+During normal operation of an OpenShift cluster, as worker node count and
+workloads increase, it becomes necessary to increase resources available to the
+control plane nodes.
+
+During SRE operations of the managed OpenShift product, there was already a
+trigger indentified, when these increases are required. Instead of requiring the
+adjustments to be performed by hand, the operator should be able to
+automatically determine if a control plane node size change is required.
+
+On the other hand, it should also be possible to determine that control plane
+nodes have become too large for a cluster, because load has decreased. In this
+case the operator should also be able to automatically decide to scale the
+control plane down again.
+
+With the automatic increase of the control plane size, cluster performance can
+remain high, while automated decrease of the size will ensure that spend for
+users in a cloud environment is only has high as needed.
+
+### User Stories
+
+* As an OpenShift administrator, I want to know that the platform will always
+  have enough resources available to it's control plane nodes without manual
+  intervention.
+* As an OpenShift administrator, I want to know that the platform can reduce
+  spending on control plane nodes if they are oversized for the current cluster
+  load.
+* As an OpenShift administrator, I want to be able to control when sizing up and
+  down occurs, to adjust it to workloads on my cluster that might be burstable
+  to prevent the automatic scaling from occurring to often.
+
+### Goals
+
+* Allow automatic vertical scaling of control plane nodes.
+* Allow users to configure the sizes that can be chosen for automatic scaling.
+* Allow users to configure the thresholds when scaling will occur for CPU and
+  memory usage.
+* Allow users to configure how often checks for automatic scaling should be run.
+* Allow users to configure how long after a scale up or scale down, no more
+  scaling should be performed.
+* Allow users to completely disable scale up.
+* Allow users to completely disable scale down.
+
+### Non-Goals
+
+* Horizontal scaling of control plane nodes
+
+## Proposal
+
+New configuration and a reconcile loop for the
```

Comment:
If this is going to be directly in the control plane machine set, then it's considered to be core functionality that we expect a significant number of users to leverage. Will this feature be built in a way that it isn't are specific and can be used on any vanilla openshift cluster without further configuration or additional operators being configured?

Any thought about whether this should be integrated into CPMS Vs being a separate add on?

Reply:
I think having an add on would be fine as well - in that case the default case could be not having it installed, so this whole feature would be opt in.
In this case SREP managed clusters could opt in using our own automation that runs after cluster installation.

Reply:
Yeah I think whether this is built in or not, having an EP for it makes sense.

CC @sub-mod for PM input, do you have any thoughts on the demand for this being built into the cluster vs being an add on?

Reply:
> Will this feature be built in a way that it isn't are specific and can be used on any vanilla openshift cluster without further configuration...

It's hard for me to imagine us knowing enough to be able to default `instancesSizes` out of the box.  But we ship [the cluster autoscaler][1] (MachineSets) and [the horizontal pod autoscaler][2] (deployments and such) baked into the OpenShift release payload.  And we ship [the custom metrics autoscaler][3] (deployments and such) as a separate operator.  So it seems like there is precedent both ways, and not a lot to decide between approaches.  One benefit to a separate operator is that OLM packages can be uninstalled, while [cluster-capabilities cannot be][4].  So if we expect some very-resource-concious consumers to not need this particular functionailty, that might be enough of an argument to push it into a separate operator?

[1]: https://docs.openshift.com/container-platform/4.13/machine_management/applying-autoscaling.html
[2]: https://docs.openshift.com/container-platform/4.13/nodes/pods/nodes-pods-autoscaling.html
[3]: https://docs.openshift.com/container-platform/4.13/nodes/cma/nodes-cma-autoscaling-custom.html
[4]: https://github.com/openshift/enhancements/blob/da2ef4e0702751c3f5f7faa393282b5dd823195e/enhancements/installer/component-selection.md#capabilities-cannot-be-uninstalled

Reply:
I had a conversation with my team at standup yesterday in length about this topic and we all came to the conclusion that having this initially as SD owned and then trying to deliver via OLM for those who want to have it in regular openshift clusters would be our preferred path forward.

I think that aligns with where others landed on this conversation too